crate/cli/src/actions/findex/index.rs

```diff
@@ -0,0 +1,136 @@
+use std::collections::{HashMap, HashSet};
+
+use clap::Parser;
+use cloudproof_findex::{
+    db_interfaces::DbInterfaceError,
+    reexport::{
+        cosmian_crypto_core::FixedSizeCBytes,
+        cosmian_findex::{Data, IndexedValue, IndexedValueToKeywordsMap, Keyword, Label, UserKey},
+    },
+    Configuration, InstantiatedFindex,
+};
+use cosmian_findex_client::FindexClient;
+use serde::{Deserialize, Serialize};
+use tracing::trace;
+
+use super::FindexParameters;
+use crate::{
+    actions::console,
+    error::result::{CliResult, CliResultHelper},
+};
+
+// to be deleted - start
+// todo(manu): replace this by adding a CLI argument to provide the dataset file
+
+#[allow(non_snake_case)]
+#[derive(Debug, Deserialize, Serialize)]
+pub struct User {
+    pub(crate) firstName: String,
+    pub(crate) lastName: String,
+    pub(crate) phone: String,
+    pub(crate) email: String,
+    pub(crate) country: String,
+    pub(crate) region: String,
+    pub(crate) employeeNumber: String,
+    pub(crate) security: String,
+}
+
+impl User {
+    #[must_use]
+    pub fn values(&self) -> Vec<String> {
+        vec![
+            self.firstName.clone(),
+            self.lastName.clone(),
+            self.phone.clone(),
+            self.email.clone(),
+            self.country.clone(),
+            self.region.clone(),
+            self.employeeNumber.clone(),
+            self.security.clone(),
+        ]
+    }
+}
+
+/// Get the users from the dataset
+/// # Errors
+/// It returns an error if the dataset cannot be read or if the dataset cannot
+/// be deserialized into a list of users
+pub fn get_users() -> Result<Vec<User>, DbInterfaceError> {
+    trace!("Current working directory: {:?}", std::env::current_dir()?);
+    let dataset = std::fs::read_to_string("../../crate/client/datasets/users.json")?;
+    serde_json::from_str::<Vec<User>>(&dataset)
+        .map_err(|e| DbInterfaceError::Serialization(e.to_string()))
+}
+// to be deleted - end
+
+/// Index data with Findex
+#[derive(Parser, Debug)]
+#[clap(verbatim_doc_comment)]
+pub struct IndexAction {
+    #[clap(flatten)]
+    pub findex_parameters: FindexParameters,
+}
+
+impl IndexAction {
+    /// Process the server version action.
+    ///
+    /// # Arguments
+    ///
+    /// * `findex_rest_client` - The Findex server client instance used to
+    ///   communicate with the Findex server server.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the version query fails or if there is an issue
+    /// writing to the console.
+    #[allow(clippy::future_not_send)] // todo(manu): remove this
+    pub async fn process(&self, findex_rest_client: &FindexClient) -> CliResult<()> {
+        let config = Configuration::Rest(
+            findex_rest_client.client.clone(),
+            findex_rest_client.server_url.clone(),
+            findex_rest_client.server_url.clone(),
+        );
```

Comment:
This seems strange to me.

If I follow you well, upon calling a CLI command, you call this function process with the client object (idk what it is). But since I doubt you serialize and store this client upon returning from the CLI, it means it is instantiated each time the CLI is called. Moreover, it seems to me that only one Findex operation can be performed per CLI invocation. Therefore I conclude that this clone is useless since the caller of `process` will not reuse this `FindexClient`.

Please, point to me where my reasoning is flawed.

More generally, I do not understand where the necessity of this `FindexClient` object and this action mechanism comes from. Imho, the ideally simplistic flow is:

1. user invokes the CLI with some arguments
2. the CLI parses its conf/ENV variables to findex the server URL. If it cannot be found, it becomes a *required* argument.
3. the CLI parses the provided arguments, returning an error upon reading an invalid one. There are 3 (or 4) possible commands: `search`, `insert`, `delete` (additionally `compact`). The validity of the rest of the argument depends on the identity of the command, and thus the main body of the CLI is just a `match` on the parsed command.
4. after parsing all arguments, if no error was thrown, then it means the parsed command will be performed: the CLI therefore instantiates a Findex object.
5. The CLI runs the appropriate Findex command with its arguments, and returns the result to the user.

As I said, this is rather simplistic and may need to be developed, for example in order to add a non-Findex command like it seems you are planning to. However, I don't think we need dedicated types with their own `proceed` functions and to pass around some `FindexClient`  object.

Reply:
The `FindexClient` is just a HTTPS Client which is initialized once and then pass to `clap` actions. About this clone, your reasoning is correct, I've removed the shared reference and passed directly the object to be consumed by actions. Yes the CLI instantiates a HTTPS client each time and does not chain Findex operations (that's what we want).
About the CLI flow, since the 4 operations `search`, `add`, `delete` and `compact` have different arguments, the 4 corresponding `process` functions (which are basic functions) are used: 
- to parse those arguments
- to give them properly to Findex API
- to display to stdout the Findex output